### ISSUE TITLE
Angular 12 ng build --watch won't update running ionic serve application

### ISSUE THREAD

**Issue by R4yYx** [angular/angular-cli#20962]
# 🐞 Bug report

- [ ] new
- [x] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] extract-i18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

 Yes it used to work with @angular-devkit/build-angular: 0.1102.13

### Description

 Building an Angular lib with the newest version (ng build), doesn't update the ionic application using this lib, run with ionic serve. With bigger libs the ng build process can outrun the serve or gets stuck. 

## 🔬 Minimal Reproduction

Can be reproduced using this [bug repo](https://github.com/R4yYx/Angular12Repo). First run ng build --watch in the lib folder, then ionic serve in the app folder. Changing anything in the lib component back-button won't update it in the application, even though both app and lib recompile.

To change to the working version run `npm run goodvers`
Updating anything now will update it in the browser like it is supposed to be.

**Comment by hardmaster92** [comment]
I'm not sure if Ionic support Angular 12 yet

**Comment by alan-agius4** [comment]
This is expected because `node_modules` are considered immutable in Webpack 5 unless the "version" changes.

See: 
https://webpack.js.org/configuration/other-options/#cachemanagedpaths
https://webpack.js.org/configuration/other-options/#immutablepaths

**Comment by keo9ren** [comment]
Is there any "straighforward workaround" for this (having a app and a lib and develop using npm link)?

1. As I understood I could go and change the version number of the lib every time I want to see the  change reflected in the app, but the dev ux wouldn't be that great.
2. Move to a monorepo
3.  Maybe tweak the webpack config of the angular app, with the options from the post above, but how exactly?
4. Any other alternatives?

Is a development approach like this even considered "good practice" by the angular devs?

Any thoughts on this are highly appreciated. Thanks.

**Comment by keo9ren** [comment]
Solved it by setting NG_BUILD_CACHE Env Variable to 0. 

For Powershell Users:
$env:NG_BUILD_CACHE=0

**Comment by restfulhead** [comment]
Just ran into this as well. I'm not using Ionic, just another angular app that's linking to a library. Setting `NG_BUILD_CACHE` to 0 works, but the consuming app now seems to compile twice and it's really slow. :/

**Comment by kruczjak** [comment]
Same here, we're using library which was also created with angular cli.
Everything was working great when running `ng build --watch` in library and symlinking it to application with `npm link`. However after ugrading to Angular 12 only app is recompiling itself when `NG_BUILD_CACHE` is set to `0` and also very, very slowly.

**Comment by kruczjak** [comment]
Following https://github.com/angular/angular-cli/issues/20962#issuecomment-849579850 I've made a simple patch into build-angular: https://github.com/kruczjak/angular-cli/commit/955d2d388efeb93862162837e3a64834be219bdc

After this change using `NG_BUILD_CACHE=0` is no longer needed and rebuilds are much faster (for me it was `65s` -> `11s`). Maybe we could get it configurable in build options?

**Comment by restfulhead** [comment]
@alan-agius4 Would it be possible to re-open this issue and consider @kruczjak 's patch?

**Comment by alan-agius4** [comment]
Let me reopen this to discuss the issue in our next team meeting if we should provide an option to disable managed paths.

In general, if this would will be added it would most likely be configured via an environment variable.

That being said we don’t really recommend using npm linking.

**Comment by kruczjak** [comment]
@alan-agius4 Using `npm link` is suggested here: https://angular.io/guide/creating-libraries#linked-libraries 
Is there any different way to achieve similar effect?

**Comment by alan-agius4** [comment]
That is probably from the time were Angular CLI didn’t have a monorepo/multi-project concept were libraries and apps can live side my side in the same got repo.

An alternative to npm linking can be using TypeScript configuration path mappings.

```js
{
  "compilerOptions": {
    // ...
    // paths are relative to `baseUrl` path.
    "paths": {
      "@mylib/*": [
       “../../my-lib/dist/*",
       “node_modules/my-lib/*
      ]
    }
  }
}
```

**Comment by dgp1130** [comment]
Discussed this in the CLI triage meeting today and it's a bit awkward since Webpack caching uses the package version to identify changes in `node_modules/`. They get around this problem by exposing an option, but we'd rather not add Webpack-specific options on Angular and we also don't want to ask users to remember and use such an option as we'd rather this "just worked" by default, before the user wastes hours debugging stale builds.

Our main thought to solve this problem is to "stamp" library builds by adding a unique string to the end of the `package.json` version. This would create a version like: `0.1.2+0123456789`, which would change on each build either using a hash of the entire package or a simple timestamp of the build. This causes the package version to change on every build of the library so Webpack knows to invalidate its cache.

The tricky question is when to stamp the version. Our initial approach is to simply stamp the build when the user runs `ng build --watch` on a library. A user would only ever do this for a development build, so we can be reasonably sure that the library is being linked to some application somewhere, meaning the stamped version will be needed to bust the Webpack cache of that application. If you only `ng build` a library, that is treated as a production build where the user would *not* want their version stamped, so we won't modify it in this case.

There is an edge case if the user already has a `0.1.2+something-else` string in their version. I'm not able to immediately confirm the semantics, but from my understanding anything after the `+` is meant to identify the build, so if the user already has this then Angular should probably leave it alone, since that should be sufficient for caching unless they are misusing `+` semantics (which is possible and I could be wrong here, we should probably double check).

This does break one expected invariant about the CLI, which is that the output of `ng build` and `ng build --watch` are not the same. This is a bit awkward and doesn't make *logical* sense, but this difference in behavior allows us to more closely align with what the user wanted, which we think is a good trade off in this case. It also means that if a user does a one-off build of their library, it won't bust the Webpack cache of their application. This is an unfortunate sharp edge, but we don't expect this to be a common use case to actually affect many users.

TL;DR: We will make `ng build --watch` "stamp" the build by appending a timestamp / hash to the package version to bust the Webpack cache automatically. `ng build` *without* `--watch` will **not** modify the version at all and keep its current behavior.

For future reference, one alternative we discussed was adding a new `stampVersion` flag with the above behavior of appending a hash / timestamp to the output `package.json`. We would then migrate existing apps (and update `ng new`) to enable this by default in their development configuration (but would be `false` for the default production configuration). By doing this, dev builds would always include the cache busting string while production builds would leave the package version unchanged.

This sounds great but introduces a lot more complexity. `stampVersion` would need to be public API due to how flexible Angular configurations are, which adds a lot more configurability but is also much more effort to support and maintain. Migrating existing apps is also tricky because Angular configurations are so powerful. Users may have multiple configurations which they see as "development" and want to add this flag to, and we don't have an understanding of this and can't intelligently add this flag automatically for users. Such a migration is effectively limited to the configuration named `development`, since we generate it by default and it has common (though not universal) semantics.

Our thought process for now it to prefer the `ng build --watch` solution has it does not affect the public API and dodges the configurations issue.

**Comment by SchnWalter** [comment]
Sorry for the noise, but I just wanted to drop by and say:

> Wow, thanks for figuring this out and for providing a quick fix!

**Comment by alan-agius4** [comment]
ng-packagr version 12.0.6+ contains the version stamping fix that @dgp1130 described above.
